File: mobile/lib/models/enums.dart
/// Enumerations of contract v1 (docs/architecture.md, section 5).
///
/// Every parser keeps an `unknown` value so that a newer backend adding a
/// value never crashes the app: the contract says unknown fields and values
/// are ignored, not rejected.
library;

/// `context`: an SSH login or a sudo invocation.
enum RequestContext {
  ssh('ssh'),
  sudo('sudo');

  const RequestContext(this.wire);

  /// The string sent and received on the wire.
  final String wire;

  static RequestContext parse(String? value) {
    if (value == 'sudo') return RequestContext.sudo;
    return RequestContext.ssh;
  }
}

/// Request `status`.
enum RequestStatus {
  pending('pending'),
  approved('approved'),
  denied('denied'),
  timeout('timeout'),
  whitelisted('whitelisted'),
  blockedIp('blocked_ip'),
  blockedGeo('blocked_geo'),
  notified('notified'),
  unknown('unknown');

  const RequestStatus(this.wire);

  final String wire;

  static RequestStatus parse(String? value) {
    for (final status in RequestStatus.values) {
      if (status.wire == value) return status;
    }
    return RequestStatus.unknown;
  }

  /// True when the login or command was let through.
  bool get isAllowed =>
      this == approved || this == whitelisted || this == notified;

  /// True when the login or command was refused.
  bool get isRefused =>
      this == denied || this == timeout || this == blockedIp || this == blockedGeo;
}

/// `decided_by`: what produced the outcome of a request.
enum DecidedBy {
  admin('admin'),
  whitelist('whitelist'),
  timeout('timeout'),
  autoblock('autoblock'),
  georule('georule'),
  unknown('unknown');

  const DecidedBy(this.wire);

  final String wire;

  static DecidedBy? parse(String? value) {
    if (value == null) return null;
    for (final by in DecidedBy.values) {
      if (by.wire == value) return by;
    }
    return DecidedBy.unknown;
  }
}

/// `verdict` sent by the app in `POST /verdict`.
enum Verdict {
  approve('approve'),
  deny('deny'),
  approveAlways('approve_always');

  const Verdict(this.wire);

  final String wire;
}

/// Push `type`.
enum PushType {
  accessRequest('access_request'),
  accessNotice('access_notice'),
  requestDecided('request_decided'),
  unknown('unknown');

  const PushType(this.wire);

  final String wire;

  static PushType parse(String? value) {
    for (final type in PushType.values) {
      if (type.wire == value) return type;
    }
    return PushType.unknown;
  }
}

File: backend/internal/model/enums.go
// Package model holds the types shared by the handlers, the rules engine and
// the database layer: the enumerations of the contract, the database rows and
// the JSON shapes sent to the agent and the phones.
package model

// Contexts of a request.
const (
	ContextSSH  = "ssh"
	ContextSudo = "sudo"
)

// Modes of a request.
const (
	ModeEnforce = "enforce"
	ModeNotify  = "notify"
)

// Request statuses.
const (
	StatusPending     = "pending"
	StatusApproved    = "approved"
	StatusDenied      = "denied"
	StatusTimeout     = "timeout"
	StatusWhitelisted = "whitelisted"
	StatusBlockedIP   = "blocked_ip"
	StatusBlockedGeo  = "blocked_geo"
	StatusNotified    = "notified"
)

// Values of requests.decided_by.
const (
	DecidedByAdmin     = "admin"
	DecidedByWhitelist = "whitelist"
	DecidedByTimeout   = "timeout"
	DecidedByAutoblock = "autoblock"
	DecidedByGeoRule   = "georule"
)

// Verdicts sent by the app.
const (
	VerdictApprove       = "approve"
	VerdictDeny          = "deny"
	VerdictApproveAlways = "approve_always"
)

// Reasons returned to the agent.
const (
	ReasonAdmin      = "admin"
	ReasonWhitelist  = "whitelist"
	ReasonTimeout    = "timeout"
	ReasonBlockedIP  = "blocked_ip"
	ReasonBlockedGeo = "blocked_geo"
	ReasonNotify     = "notify"
)

// Push message types.
const (
	PushAccessRequest  = "access_request"
	PushAccessNotice   = "access_notice"
	PushRequestDecided = "request_decided"
)

// Device platforms.
const (
	PlatformAndroid = "android"
	PlatformIOS     = "ios"
)

// Server operating systems.
const (
	OSLinux   = "linux"
	OSWindows = "windows"
	OSDarwin  = "darwin"
)

// ValidContext reports whether s is a context of the contract.
func ValidContext(s string) bool { return s == ContextSSH || s == ContextSudo }

// ValidMode reports whether s is a mode of the contract.
func ValidMode(s string) bool { return s == ModeEnforce || s == ModeNotify }

// ValidStatus reports whether s is a request status of the contract.
func ValidStatus(s string) bool {
	switch s {
	case StatusPending, StatusApproved, StatusDenied, StatusTimeout,
		StatusWhitelisted, StatusBlockedIP, StatusBlockedGeo, StatusNotified:
		return true
	}
	return false
}

// ValidPlatform reports whether s is a device platform of the contract.
func ValidPlatform(s string) bool { return s == PlatformAndroid || s == PlatformIOS }

// ValidOS reports whether s is a server operating system of the schema.
func ValidOS(s string) bool { return s == OSLinux || s == OSWindows || s == OSDarwin }

// VerdictFor maps a final request status to the verdict and reason returned
// to the agent. Pending has no verdict yet; the caller must not ask for it.
func VerdictFor(status string) (verdict, reason string) {
	switch status {
	case StatusApproved:
		return VerdictApprove, ReasonAdmin
	case StatusDenied:
		return VerdictDeny, ReasonAdmin
	case StatusWhitelisted:
		return VerdictApprove, ReasonWhitelist
	case StatusNotified:
		return VerdictApprove, ReasonNotify
	case StatusBlockedIP:
		return VerdictDeny, ReasonBlockedIP
	case StatusBlockedGeo:
		return VerdictDeny, ReasonBlockedGeo
	default: // timeout, and anything unexpected: fail closed
		return VerdictDeny, ReasonTimeout
	}
}
